Inline hooks in ivalue::Future

lw opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#57354 Inline hooks in ivalue::Future**
* #57353 Use Devices instead of DeviceIndexes in Future
* #57294 Use Devices instead of DeviceIndexes in TensorPipe agent
* #57345 Rename DeviceGuardImplInteface's getStreamFromPool method
* #57293 Add device() method to c10::Event
* #57292 Add pybind type caster for c10::Device

The ivalue::Future class used to have some hooks, defined as separate protected virtual methods, so that they could be overridden by the CUDAFuture subclass. Now that CUDAFuture has been merged into ivalue::Future those hooks can be "inlined" to where they're used, hopefully making the code more readable as it puts related things closer together.

Differential Revision: [D28117199](https://our.internmc.facebook.com/intern/diff/D28117199/)

lw committed:
Inline hooks in ivalue::Future

The ivalue::Future class used to have some hooks, defined as separate protected virtual methods, so that they could be overridden by the CUDAFuture subclass. Now that CUDAFuture has been merged into ivalue::Future those hooks can be "inlined" to where they're used, hopefully making the code more readable as it puts related things closer together.

Differential Revision: [D28117199](https://our.internmc.facebook.com/intern/diff/D28117199/)

[ghstack-poisoned]

lw commented on aten/src/ATen/core/ivalue_inl.h:
```diff
@@ -343,37 +343,29 @@ struct C10_EXPORT ivalue::Future final : c10::intrusive_ptr_target {
 
-    if (!eptr_) {
-      postWaitHook(value_);
-    }
+    synchronizeWithCurrentStreams();
   }
 
   /**
    * Wait on the future until it completes and throw an
    * exception if an error exists.
    */
   void waitAndThrow() {
-    std::unique_lock<std::mutex> lock(mutex_);
-    while (!completed_) {
-      finished_cv_.wait(lock);
-    }
+    wait();
 
     if (eptr_) {
       std::rethrow_exception(eptr_);
     }
-
-    postWaitHook(value_);
```
One assumption needed to understand why this simplification works is that, when `eptr_ != nullptr` we have no events/dataptrs stored, hence `synchronizeWithCurrentStreams` is a no-op.

mrshenli commented on aten/src/ATen/core/ivalue_inl.h:
```diff
@@ -343,37 +343,29 @@ struct C10_EXPORT ivalue::Future final : c10::intrusive_ptr_target {
-   * efficient.
+   * the storage pointers for all tensors in IValue can be passed as well. These
+   * DataPtrs are used to synchronize CUDA streams. If data_ptrs isn't given we
+   * will attempt to extract it from the value, if we need. Thus one only needs
```
> If data_ptrs isn't given we will attempt to extract it from the value, if we need. 

Shall we also explain when "if we need" == true?

mrshenli commented on aten/src/ATen/core/ivalue_inl.h:
```diff
@@ -385,25 +377,46 @@ struct C10_EXPORT ivalue::Future final : c10::intrusive_ptr_target {
     value_ = std::move(value);
     completed_ = true;
 
+    currentDevice_ = impl_.getDevice();
```
I guess the main benefits of doing this here is that we can now avoid wrapping the callbacks?

lw replied:
My main reason was to avoid too much branching (i.e., all those `if (impl_ == nullptr)` checks) which adds complexity to the code, hurts readability and makes testing less effective (more paths and combinations of paths to cover).

mrshenli reviewed (approved):
LGTM. This is good simplification.

lw committed:
Update on "Inline hooks in ivalue::Future"


The ivalue::Future class used to have some hooks, defined as separate protected virtual methods, so that they could be overridden by the CUDAFuture subclass. Now that CUDAFuture has been merged into ivalue::Future those hooks can be "inlined" to where they're used, hopefully making the code more readable as it puts related things closer together.

Differential Revision: [D28117199](https://our.internmc.facebook.com/intern/diff/D28117199/)

[ghstack-poisoned]

lw committed:
Update on "Inline hooks in ivalue::Future"


The ivalue::Future class used to have some hooks, defined as separate protected virtual methods, so that they could be overridden by the CUDAFuture subclass. Now that CUDAFuture has been merged into ivalue::Future those hooks can be "inlined" to where they're used, hopefully making the code more readable as it puts related things closer together.

Differential Revision: [D28117199](https://our.internmc.facebook.com/intern/diff/D28117199/)

[ghstack-poisoned]

lw committed:
Update on "Inline hooks in ivalue::Future"


The ivalue::Future class used to have some hooks, defined as separate protected virtual methods, so that they could be overridden by the CUDAFuture subclass. Now that CUDAFuture has been merged into ivalue::Future those hooks can be "inlined" to where they're used, hopefully making the code more readable as it puts related things closer together.

Differential Revision: [D28117199](https://our.internmc.facebook.com/intern/diff/D28117199/)

[ghstack-poisoned]

lw committed:
Update on "Inline hooks in ivalue::Future"


The ivalue::Future class used to have some hooks, defined as separate protected virtual methods, so that they could be overridden by the CUDAFuture subclass. Now that CUDAFuture has been merged into ivalue::Future those hooks can be "inlined" to where they're used, hopefully making the code more readable as it puts related things closer together.

Differential Revision: [D28117199](https://our.internmc.facebook.com/intern/diff/D28117199/)

[ghstack-poisoned]

lw committed:
Update on "Inline hooks in ivalue::Future"


The ivalue::Future class used to have some hooks, defined as separate protected virtual methods, so that they could be overridden by the CUDAFuture subclass. Now that CUDAFuture has been merged into ivalue::Future those hooks can be "inlined" to where they're used, hopefully making the code more readable as it puts related things closer together.

Differential Revision: [D28117199](https://our.internmc.facebook.com/intern/diff/D28117199/)

[ghstack-poisoned]

lw committed:
Update on "Inline hooks in ivalue::Future"


The ivalue::Future class used to have some hooks, defined as separate protected virtual methods, so that they could be overridden by the CUDAFuture subclass. Now that CUDAFuture has been merged into ivalue::Future those hooks can be "inlined" to where they're used, hopefully making the code more readable as it puts related things closer together.

Differential Revision: [D28117199](https://our.internmc.facebook.com/intern/diff/D28117199/)

[ghstack-poisoned]
